pkg/controller/common/reconciler.go

```diff
@@ -198,15 +203,47 @@ func (r *BaseIngressReconciler) reconcileRoute(ctx context.Context, ci networkin
 }
 
 func (r *BaseIngressReconciler) reconcileDeletion(ctx context.Context, ci networkingv1alpha1.IngressAccessor) error {
-	// TODO: something with a finalizer?  We're using owner refs for
-	// now, but really shouldn't be using owner refs from
-	// cluster-scoped ClusterIngress to a namespace-scoped K8s
-	// Service.
-	return nil
+	// get list of ingress object for a namespace
+	ingressList := networkingv1alpha1.IngressList{}
+	if err := r.Client.List(ctx, &client.ListOptions{
+		Namespace: ci.GetNamespace(),
+	}, &ingressList); err != nil {
+		return err
+	}
+	// If particular namespace has only one ingress object then after deletion namespace will have empty ingress object
+	// So remove namespace from SMMR
+	if len(ingressList.Items) == 1 {
+		for i := range ingressList.Items {
+			// In order to double check that we are reconciling proper ingress check with name and namespace
+			if ci.GetNamespace() == ingressList.Items[i].Namespace && ci.GetName() == ingressList.Items[i].Name {
+				smmr := &maistrav1.ServiceMeshMemberRoll{}
+				if err := r.Client.Get(ctx, types.NamespacedName{Name: smmrName, Namespace: smmrNamespace}, smmr); err != nil {
+					return err
+				}
+				for i, val := range smmr.Spec.Members {
+					if val == ci.GetNamespace() {
+						smmr.Spec.Members = append(smmr.Spec.Members[:i], smmr.Spec.Members[i+1:]...)
+						break
+					}
+				}
+				if err := r.Client.Update(ctx, smmr); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	for i := range ci.GetFinalizers() {
```

Comment:
We shouldn‘t remove this from anywhere in the list. Rather we should check before even doing something if „our“ finalizer is the first. Only then will we do something. At the end we can then just drop the first element. Please check the reference I left above.